server/src/main/java/org/opensearch/action/search/TransportSearchAction.java

```diff
@@ -431,49 +440,68 @@ private void executeRequest(
         if (originalSearchRequest.isPhaseTook() == null) {
             originalSearchRequest.setPhaseTook(clusterService.getClusterSettings().get(SEARCH_PHASE_TOOK_ENABLED));
         }
-        SearchRequestOperationsListener.CompositeListener requestOperationsListeners = searchRequestOperationsCompositeListenerFactory
-            .buildCompositeListener(originalSearchRequest, logger);
-        SearchRequestContext searchRequestContext = new SearchRequestContext(requestOperationsListeners, originalSearchRequest);
-        searchRequestContext.getSearchRequestOperationsListener().onRequestStart(searchRequestContext);
-
-        PipelinedRequest searchRequest;
-        ActionListener<SearchResponse> listener;
-        try {
-            searchRequest = searchPipelineService.resolvePipeline(originalSearchRequest);
-            listener = searchRequest.transformResponseListener(originalListener);
-        } catch (Exception e) {
-            originalListener.onFailure(e);
-            return;
-        }
 
-        ActionListener<SearchRequest> requestTransformListener = ActionListener.wrap(sr -> {
-            if (searchQueryMetricsEnabled) {
-                try {
-                    searchQueryCategorizer.categorize(sr.source());
-                } catch (Exception e) {
-                    logger.error("Error while trying to categorize the query.", e);
-                }
-            }
+        final Span requestSpan = tracer.startSpan(SpanBuilder.from(task, actionName));
+        try (final SpanScope spanScope = tracer.withSpanInScope(requestSpan)) {
 
-            ActionListener<SearchSourceBuilder> rewriteListener = buildRewriteListener(
-                sr,
-                task,
-                timeProvider,
-                searchAsyncActionProvider,
-                listener,
-                searchRequestContext
-            );
-            if (sr.source() == null) {
-                rewriteListener.onResponse(sr.source());
+            SearchRequestOperationsListener.CompositeListener requestOperationsListeners;
+            if (tracer.isRecording()) {
+                originalListener = TraceableActionListener.create(originalListener, requestSpan, tracer);
```

Comment:
You're still re-assigning a method argument here. You'll want to structure this something like:

```
final ActionListener<SearchResponse> updatedListener;
if (tracer.isRecording()) {
    updatedListener = TraceableActionLister...
    ...
 } else {
     updatedListener = originalListener;
    ...
}
```

and then use `updatedListener` throughout the rest of the method. Sometimes you can avoid this awkwardness by refactoring things out to separate methods too, not sure if that's feasible here.

Reply:
I see the issue since other parts of the code may rely on the unwrapped `originalListener`. `TraceableActionListener.create(...)` already checks `tracer.isRecording()` so we can simplify to just
```
updatedListener = TraceableActionListener.create(...)
```